### PR TITLE
TheiaValidate_PHB: new features and new Docker image from TheiaValidate repository

### LINKED ISSUE #181: Improve TheiaValidate
:cool:

### :pushpin: Explain the Request

TheiaValidate is cool, but it could be cooler. 

### :books: Context

We've come across several requests/realizations on how to improve this workflow as it has been used more. 

### :chart_with_upwards_trend: Desired Behavior

- [ ] Create an optional input file that matches columns to columns with different names. For example:
  |table1|table2|
  |------|--------|
  |genome_length|assembly_length|
  |ts_mlst|MLST_results|
  |amrfinder_genes|amr_genes|
- [ ] Include only items in the difference table that do not meet validation criteria. This is because it becomes increasingly difficult to figure out what items do not actually meet the validation criteria when it only outputs exact matches -- for example, when doing set comparisons. 
- [ ] Include the ability to compare file contents. Not yet sure how to do this yet, but the easiest might just be see if file contents are identical.
- [ ] Enable the ability to compare three data tables? Not totally sure how this would work though, unless we set one table as the "true" table and compared the other two to that one. This would be a much heavier lift.

### :information_source: Additional Information

As more people use TheiaValidate, more requests will likely be added to this list.

### LINKED ISSUE #281: Move TheiaValidate into its own repository
:cool:

### :pushpin: Explain the Request

There is a lot of code in this task and it should be moved into its own repository.

### :books: Context

Reduction of tasks with hundreds of lines of code.

### PR DESCRIPTION
**Pull request by sage-wright**
~theiagen/theiavalidate#1 partially~

closes #281 

## :hammer_and_wrench: Changes Being Made

This PR moves the entirety of the TheiaValidate code into [its own repository](https://github.com/theiagen/theiavalidate). It also implements a few new features:
- a table outputting validation-criteria failures only
- a new input that can translate different column names between tables to enable comparison

### Impacted Workflows/Tasks

- TheiaValidate_PHB
- the `task compare_two_tsvs` in task_validate.wdl is now `task theiavalidate`

## :brain: Context and Rationale

The Python code was exceeding our limit of 100 lines in a task file so we decided to transfer all the code into its own repository. Also, some changes and enhancements had been requested highly and so were implemented as well.

### Inputs

Renamed inputs:
- `table1` -> `table1_name`
- `table2` -> `table2_name`

New outputs:
- `Boolean debug_output` - set to true for additional outputs
- `na_values` - provide a comma-separated list of values to be used as "NAs"
- `column_translation_tsv` - a TSV file that matches columns of different names to each other; it "translates" column names between tables

### Outputs

Renamed outputs: 
- `validation_report` -> `theiavalidate_report`
- `validation_status` -> `theiavalidate_status`
- `input_table1` -> `theiavalidate_filtered_table1`
- `input_table2` -> `theiavalidate_filtered_table2`
- `validation_differences_table` -> `theiavalidate_exact_differences`
- `theiavalidate_version` -> `theiavalidate_wf_version

New outputs:
- `theiavalidate_version` - version of TheiaValidate Python script in Docker
- `theiavalidate_criteria_differences` - same as exact_differences but only for differences that fail to meet the validation criteria

### Impacted Outputs

Several old TheiaValidate outputs have been deprecated, but are still accessible under a different name (see above).

### Terra

See: https://app.terra.bio/#workspaces/theiagen-validations/Theiagen_Wright_PHBG_Sandbox/job_history/cbb69629-f456-4e17-91ad-946397a39161
(the failure was expected)

### Scenarios  for Reviewer to Test

- compare validation outputs to outputs generated previously
- ensure values that passed validation criteria are not present in the validation criteria differences output

## :microscope: Quality checks

Pull Request (PR) checklist:
- [x] Include a description of what is in this pull request in this message.
- [x] The workflow/task has been tested locally and on Terra
- [x] The CI/CD has been adjusted and tests are passing
- [x] Everything follows the [style guide](https://theiagen.notion.site/Style-Guide-WDL-Workflow-Development-bb456f34322d4f4db699d4029050481c)

### CODE REVIEW

**Comment by michellescribner** [comment]
Tested successfully: https://app.terra.bio/#workspaces/theiagen-validations/Theiagen_Scribner_Sandbox/job_history/db64e855-159d-4d60-96e1-bb3820006b70
Tested on 40 samples that were run with 2 different GAMBIT databases. 2 samples also had small assembly length differences.
- workflow completed successfully
- differences in exact differences file matched previous results
- differences that passed validation criteria were not listed in the validation criteria differences file
Looks great! Happy to approve pending merge conflict resolution

**Review by michellescribner — APPROVED** [review]
Great job Sage!

File: tasks/utilities/task_validate.wdl
Comment (kevinlibuit) [inline tasks/utilities/task_validate.wdl:93]: Any chance we can get a [theiavdalite ](https://github.com/theiagen/theiavalidate/tree/main) v1 release (and associated docker) to use as default before merging this? And, since it's a python package, I'd suggest @cimendes as a reviewer before releasing things.

File: tasks/utilities/task_validate.wdl
Comment (kevinlibuit) [inline tasks/utilities/task_validate.wdl:66]: I may have missed this in a previous sync, but do you mind explaining the utility of forcing values to NA?

File: tasks/utilities/task_validate.wdl
Comment (kevinlibuit) [inline tasks/utilities/task_validate.wdl:65]: Oh, great call in adding a `column_translation_tsv` input! Do you have an example of this option utilized on Terra?

File: tasks/utilities/task_validate.wdl
Comment (kevinlibuit) [inline tasks/utilities/task_validate.wdl:67]: Is this for additional logging outputs?

File: tasks/utilities/task_validate.wdl
Comment (michellescribner) [reply tasks/utilities/task_validate.wdl:65]: @kevinlibuit @sage-wright I did not test this option but I can do that now

File: tasks/utilities/task_validate.wdl
Comment (sage-wright) [reply tasks/utilities/task_validate.wdl:66]: We ran into this issue when we were making a release and "NA" values were being converted to None when we actually wanted them to be "NA" and that caused a lot of problems. I think we came up with this solution together.

File: tasks/utilities/task_validate.wdl
Comment (michellescribner) [reply tasks/utilities/task_validate.wdl:65]: Tested: https://app.terra.bio/#workspaces/theiagen-validations/Theiagen_Scribner_Sandbox/job_history/45a03c16-8e84-46af-ab8b-759041b85426
Looks good though you can't have only one row in column_translation_tsv - will break the workflow. This is described in TheiaValidate repo

File: tasks/utilities/task_validate.wdl
Comment (sage-wright) [reply tasks/utilities/task_validate.wdl:93]: I made a theiavalidate 0.0.1 release and associated docker 👍

**Pull request merged**
